Fixes for cdef subclasses of int/complex builtin types

dalcinl opened the pull request:
1. Support subclasses of `complex`
```cython
cdef class Complex(complex): pass
```

2. Fix code generation for subclasses of `int` (generated C code used `PyIntObject` instead of `PyLongObject`)
```cython
cdef class Int(int): pass
```

da-woods commented:
`PyLongObject` shouldn't be allowed currently because it's a `PyVarObject` so isn't safe to inherit from. There was [a PR to change that](https://github.com/cython/cython/pull/5212) but I was waiting for some changes to tighten up the tracking of inheritance before that.

So for now inheriting from `int` shouldn't be allowed. Essentially it fails if you have a large value and anything else in the struct.

dalcinl commented:
Yet, if the subclass is trivial, like in the example I added, there is nothing wrong with it. 
I use a @cython.final subclass of `int` in mpi4py to make a trivial wrapper storing a small special pointer-like constant to interact with a C-API. Moreover, `PyLongObject` is indeed the only member in the struct, and the generated C code properly uses 
https://github.com/mpi4py/mpi4py/blob/master/src/mpi4py/MPI.src/msgbuffer.pxi#L29.

This works perfectly fine right now, and IMHO I do not see any reason to actively break my working code. Obviously, there is a conflict of interest in my opinion.

dalcinl commented:
@da-woods Cython generates C code with `__Pyx_validate_extern_base`. 

The following type definition:
```cython
cdef class Int(int):
    cdef int dummy
````
triggers the following exception at module import time:
```
TypeError: inheritance from PyVarObject types like 'int' not currently supported
```

Therefore, I have to reiterate my request to please not break valid and safe use cases of `int` subclassing.

da-woods commented:
If `__Pyx_validate_extern_base` ends up catching it I don't want to break it for the sake of breaking it.

That probably suggests we should loosen the compile-time checks that we have for some builtin types instead to bring them in line with the external base check. (But not in this PR)

I'm happy with this PR then to go into 3.1 (I think the `int` side of it couldn't be backported though since it'd break Python 2).

dalcinl commented:
> (I think the `int` side of it couldn't be backported though since it'd break Python 2).

What do you mean? Backported to what Cython release? I believe the `int` part of it is a recent regression in cython@master. Cython's testsuite no longer runs in Python 2 due to the use of f-strings.

PS: I just moved the test I added from compile to run.

da-woods commented:
> What do you mean? Backported to what Cython release? I believe the `int` part of it is a recent regression in cython@master. Cython's testsuite no longer runs in Python 2 due to the use of f-strings.

I just meant "don't backport it to the Cython 3.0 branch". But if it's fixing a recent regression that's only in master then you wouldn't be doing that anyway.

scoder commented:
I added comments in the merge commit https://github.com/cython/cython/commit/a2120861a9c3f87b246c9ade4bed68b84ce95dba
Could you please send a new PR for those?

dalcinl commented:
@scoder I'm confused. You pushed a fix for the `elif` part, but not for your other requests. 
Do you still want the following in a PR? Or would you rather want me to just use `Builtin.complex_type` ?

```patch
diff --git a/Cython/Compiler/Nodes.py b/Cython/Compiler/Nodes.py
index eed605928..b2477df26 100644
--- a/Cython/Compiler/Nodes.py
+++ b/Cython/Compiler/Nodes.py
@@ -5485,7 +5485,7 @@ class CClassDefNode(ClassDefNode):
             if base_type in (PyrexTypes.c_int_type, PyrexTypes.c_long_type, PyrexTypes.c_float_type):
                 base_type = env.lookup(base_type.sign_and_name()).type
             elif base_type is PyrexTypes.c_double_complex_type:
-                base_type = Builtin.builtin_scope.lookup('complex').type
+                base_type = env.lookup('complex').type
 
             if base_type is None:
                 error(base.pos, "First base of '%s' is not an extension type" % self.class_name)
```
